[CI:BUILD] Combine buildahimage Containerfiles

AndASM opened the pull request:
<!--
Thanks for sending a pull request!

Please make sure you've read and understood our contributing guidelines
(https://github.com/containers/buildah/blob/main/CONTRIBUTING.md) as well as ensuring
that all your commits are signed with `git commit -s`.
-->

#### What type of PR is this?

<!--
Please label this pull request according to what type of issue you are
addressing, especially if this is a release targeted pull request.

Uncomment only one `/kind <>` line, hit enter to put that in a new line, and
remove leading whitespace from that line:
-->
/kind cleanup

#### What this PR does / why we need it:
A single `Containerfile` should be easier to maintain and more reliable than having three nearly identical files. The new file uses an argument named `RELEASE` set to one of: `stable`, `testing`, or `upstream`.

#### How to verify it
Build the containerfile with a RELEASE build parameter set. Something like:
```bash
buildah bud -t buildah:test-build --build-arg "RELEASE=upstream" contrib/buildahimage
```
#### Which issue(s) this PR fixes:
None
#### Special notes for your reviewer:
This will break the CI builder. I believe the image builder script combines the build context path and the image version into a single parameter, using basename on the path to get the version/subimage name. That would need updating to build this properly. But given how many mistakes and inconsistencies these files had, I think it may make maintenance easier.

As far as I can tell, the relevant files would be `.cirrus.yml` in this repository and https://github.com/containers/automation_images/blob/main/build-push/bin/main.sh. Since that is a multi-repository change involving the CI system, I haven't attempted it in this pull request. I'm new to such things.

#### Does this PR introduce a user-facing change?

<!--
If no, just write `None` in the release-note block below. If yes, a release note
is required: Enter your extended release note in the block below. If the PR
requires additional action from users switching to the new release, include the
string "action required".

For more information on release notes please follow the kubernetes model:
https://git.k8s.io/community/contributors/guide/release-notes.md
-->

```release-note
None
```

rhatdan commented:
LGTM
I will wait for @cevich to respond, since I think his builders will need to be modified.

rhatdan commented:
/approve

cevich commented:
> I will wait for @cevich to respond, since I think his builders will need to be modified.

Thanks Dan.  This will most certainly break the build automation.  Though the changes seem correct on quick inspection, we def. don't want to merge this until [the automation is fixed.](https://github.com/containers/automation_images/blob/main/build-push/bin/main.sh)  I'll need to find my thinking-cap to figure out a solution.

FYI- @AndASM I'm going to update the PR description with a magic keyword.  If you rebase/force-push to this PR, you'll see three image-build test tasks pop up.  They can be manually triggered to do a test-run of our automation (which will fail).  This will be helpful once I figure out how to update our build script.

cevich commented:
/hold

cevich commented on contrib/buildahimage/Containerfile:
```diff
@@ -1,40 +1,59 @@
 #
+
 FROM registry.fedoraproject.org/fedora:latest
+ARG RELEASE="${RELEASE:-stable}"
```
I'm slightly concerned the build-arg `RELEASE` is somewhat `ambiguous`.  It could be interpreted by a maintainer as meaning the buildah release version - which it most certainly is not.

I'm not so great at naming things...but would you consider something like "STAGE" or "FLAVOR" or similar?

cevich replied:
Yeah, I'm going to need to coordinate this with my build script changes.  Any objection to calling it "STAGE"?

rhatdan replied:
Stage means something in buildah.

Stick with FLAVOR, since nothing conflicts with that name.

nalind replied:
This line is equivalent to `ARG RELEASE=stable`.

AndASM replied:
Good catch nalid! I keep doing that. It's one of the language context switch glitches my brain seems to hold onto, shell vs containerfile. At least I didn't try to use any javascript, rust, or c syntax! 🤣

cevich commented on contrib/buildahimage/Containerfile:
```diff
@@ -1,40 +1,59 @@
 #
+
 FROM registry.fedoraproject.org/fedora:latest
+ARG FLAVOUR=stable
```
Lol, the *other* spelling of "flavor" is going to drive my cide-spell-checker nuts :rofl: No worries, I can roll with it.

AndASM replied:
Oh whoops! Sorry, I used English not American! Sorry sorry. I might be giving away what country I grew up in, eh!

I'll 'fix' it, since the majority here probably uses the American version. Minimize the number of people befuddled by the language!

cevich replied:
(meaning: please don't change it for my sake)

AndASM replied:
No, I think it should be changed. There is value in using a consistent dialect of the language. Predictability and all. It is very in line with my intentions with this entire pull request. Simplify maintenance, lower cognitive load, etc.

See nalind's comment about my use of shell syntax where more simple containerfile syntax could be used in your first review topic for an example of what happens. Context switching and code switching have a cognitive cost.

cevich commented:
Thanks @AndASM I'm working on updates to the build script, which we can test here shortly (assuming I can manage to spell *flavour* correctly lol).

cevich commented:
OMG :rofl: I am most definitely the befouddled one.  Too funny.  Fair enough, I'll adjust the build script to please my spell-checker lol.

AndASM commented:
> OMG 🤣 I am most definitely the befouddled one. Too funny. Fair enough, I'll adjust the build script to please my spell-checker lol.

When I decided to (finally) use my real name on my Github account so I could contribute to projects like this I created a new domain to start housing any personal coding projects or related online artifacts. I used it for my sign-off email address. I just HAD to make a joke using it. Thank you for giving me the opportunity to! 🤣

We can both be befuddled together! Everyone 'wins'!

cevich commented:
Well I certainly appreciate both the humour and the contribution :grinning: 

Build-script update: I'm working on adding test to verify the new build-arg support.  Assuming that passes, as soon as I merge https://github.com/containers/automation_images/pull/213 we can trigger the test-builds here to confirm all the planets are properly aligned, and the automation $DIETYS still look favourably on us.

TomSweeneyRedHat commented on contrib/buildahimage/Containerfile:
```diff
@@ -1,40 +1,59 @@
-    chown -R build:build /home/build
-
-ADD https://raw.githubusercontent.com/containers/buildah/main/contrib/buildahimage/stable/containers.conf /etc/containers/
+ADD ./containers.conf /etc/containers/
```
I may be misremembering, but I think this falls over in some situations.  Although far uglier, I think you should stick with:  
`ADD https://raw.githubusercontent.com/containers/buildah/main/contrib/buildahimage/stable/containers.conf /etc/containers/`
`

AndASM replied:
I would be curious on why, since that is increasing the coupling of the code to remote and external resources. That would result in ignoring local changes. Quite confusing for someone if they change the file, and it doesn't work.

TomSweeneyRedHat replied:
And it no longer applies, but these files used to be used at quay.io with a build trigger, that would fail there.  But that's no longer a consideration, just history.

AndASM replied:
Having no build context would be a special requirement wouldn't it? I guess that comes down to what the purpose of these files is. As they are, they are almost completely decoupled from this repository and heavily coupled to external services. I may be misunderstanding them.

AndASM replied:
I'm going to mark this as resolved for now. We can revisit if there are future issues. I hope that's okay!

AndASM commented:
> Then 🤞 we can try triggering the test build tasks in this PR to experience the fail^H^H^H^Hglory.

@cevich I added your changes and rebased on the latest main branch. We shall see about this glory 😆

rhatdan commented:
LGTM
@cevich PTAL

cevich commented:
I just triggered the test jobs, let's see what happens...

cevich commented:
...I see in the output, it's properly passing `--build-arg FLAVOR=stable` so that's a good sign...

cevich commented:
...great, it's building the images and manifest-list (this is going to take a while).  I'll check back in a few hours...

cevich commented:
...yay :tada:

cevich commented on contrib/buildahimage/Containerfile:
```diff
@@ -1,40 +1,59 @@
+#                   from the Fedoras Updates System.
+# FLAVOR=upstream  aquires a testing version of Buildah
+#                   from the Fedora Copr Buildsystem.
+#                   https://github.com/containers/buildah
```
```suggestion
#                   https://copr.fedorainfracloud.org/coprs/rhcontainerbot/podman-next/
```

cevich commented:
This LGTM with [one small change](https://github.com/containers/buildah/pull/4300#discussion_r990333466), and it seems github wants you to rebase anyway.  Thanks for all your work on this, I'm glad we were able to get it working!

AndASM commented:
Accepted [the change](https://github.com/containers/buildah/pull/4300#discussion_r990333466) from @cevich, and rebased to current. It looks like the automatic checks have kicked off.

Is there anything else pending for the hold on this? Let me know!

cevich commented:
Nope, we're good.  I just triggered the test-build jobs.  Assuming they go green we can merge this.  Thanks again!

cevich commented:
/lgtm
/hold
